### PR TITLE
feat(drawer): Added the drawer component to experimental

### LINKED ISSUE #1938: Introduce drawer component for react
Introduce the border component for react.  The drawer component can be seen in notification drawer as well as in other page layouts, where the drawer slide content over, in addition to the capability to lay over content.

This requires updates to:
* Accordion (https://github.com/patternfly/patternfly-next/issues/1787)
* Alert Group (https://github.com/patternfly/patternfly-next/issues/1788), and
* A Notification Icon (https://github.com/patternfly/patternfly-next/issues/1789)

Tasks:
- [ ] Accordion styling updates if applicable outside this component will be fixed with this issue.

See issue: https://github.com/patternfly/patternfly-next/issues/1786

### PR DESCRIPTION
**Pull request by dlabaj**
Fixes #1938

Added drawer component to experimental features similar to what was done for core (https://pf4.patternfly.org/components/Drawer/examples/)

TODO:

Adding unit and component tests now will update this PR with those tests.

### CODE REVIEW

**Comment by dlabaj** [comment]
> This looks good @dlabaj , but I'm a bit confused about what's different about the two examples. The first one should overlay content on the page and the second would push page content into a smaller container, right? I can't really tell from these examples if that's what's happening since it looks like the page content and the drawer are not competing for the same space.

@mcarrano The styles are messed up due to the import of pattternfly.css and base.css coming before the drawer component.  Trying to figure out how to fix it now.

**Comment by dlabaj** [comment]
@mcarrano It's fixed now can you take another look at the docs.

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import { Drawer, DrawerPanelContent } from '@patternfly/react-core/dist/esm/experimental';
+import { Alert } from '@patternfly/react-core';
+
+<Alert variant="danger" title="Warning">
+  Use this component at your own risk, it's currently under development and is subject
+  to change.
+</Alert>
+<br />
+
+## Simple Drawer Component
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md]: Sentence case her please

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    return (<React.Fragment>
+            <Button onClick={this.onClick}>Toggle Drawer</Button>
+            <Drawer isExpanded={isExpanded} panelContent={panelContent}>{drawerContent}</Drawer>
+        </React.Fragment>);
+  }
+}
+```
+
+
+## Simple Drawer Content Inline Component
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md]: Sentence case here as well

File: packages/patternfly-4/react-docs/gatsby-browser.js [hunk 3]
```diff
@@ -2,3 +2,5 @@
 // https://www.gatsbyjs.org/docs/creating-global-styles
 import './static/base.css'; // eslint-disable-line import/no-unresolved
 import '@patternfly-safe/patternfly/patternfly.css';
+// Adde experimental styles here and in eslintrc.js for the moment.  Once they have been moved out of experimental remove the CSS here and in .eslintrc.js
```
Comment (tlabaj) [inline packages/patternfly-4/react-docs/gatsby-browser.js]: Add not Adde?

File: packages/patternfly-4/react-integration/demo-app-ts/src/components/demos/DrawerDemo/DrawerDemo.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
```
Comment (tlabaj) [inline packages/patternfly-4/react-integration/demo-app-ts/src/components/demos/DrawerDemo/DrawerDemo.tsx]: Can you also update the page demo so we exercise the the new interfaces?

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md (hunk 1 above)
Comment (dlabaj) [reply packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md]: Done.

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md (hunk 2 above)
Comment (dlabaj) [reply packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md]: Done.

File: packages/patternfly-4/react-docs/gatsby-browser.js (hunk 3 above)
Comment (dlabaj) [reply packages/patternfly-4/react-docs/gatsby-browser.js]: Done.

File: packages/patternfly-4/react-integration/demo-app-ts/src/components/demos/DrawerDemo/DrawerDemo.tsx (hunk 4 above)
Comment (dlabaj) [reply packages/patternfly-4/react-integration/demo-app-ts/src/components/demos/DrawerDemo/DrawerDemo.tsx]: Added the interfaces to drawer demo.

**Comment by mcarrano** [comment]
Updates look good @dlabaj . This may be more of a CSS issue, but ion the second (inline) example, should there be some padding between the edge of the drawer and the content area? Thoughts @mcoker ?

**Comment by mcoker** [comment]
> This may be more of a CSS issue, but ion the second (inline) example, should there be some padding between the edge of the drawer and the content area?

@mcarrano that's a good point. On second thought, seems like both the main element (holds the main content) and drawer element (hold's the drawer panel's content) should have padding by default, and the option to remove it, if the user wants to add their own layout that has its own padding. What do you think of that?

**Comment by mcarrano** [comment]
Makes sense to me @mcoker . Is that something we can easily introduce?

**Comment by mcoker** [comment]
@mcarrano yep I'll go ahead and submit a PR.

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx [hunk 5]
```diff
@@ -15,7 +15,7 @@ export const DrawerPanelContent: React.SFC<DrawerPanelContentProps> = ({
   ...props
 }: DrawerPanelContentProps) => (
       <aside className={css(styles.drawerPanel, className)} {...props}>
-        <div className={css(styles.drawerPanelBody)}>
+        <div className={css(styles.drawerPanelBody, styles.modifiers.noPadding)}>
```
Comment (mcoker) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx]: looks like this just adds the modifier? It should be a prop you can define for the panel content that applies this class. Without the modifier defined (default), the panel should have padding. With the modifier applied, the panel should have no padding around it (as it does right now).

FWIW, we'll be adding the same type of prop for the drawer content element, too, but I need to make that change in core first.

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  }
+}
+```
+
+
+## Simple drawer content inline component
+
+```js
+import React {ReactFragment} from 'react';
+import { Drawer } from '@patternfly/react-core/dist/esm/experimental';
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md]: Missing `DrawerPanelContent` here, although i am wondering if we can't simplify this and only expose a Drawer component? i.e.
`<Drawer content="panel content">My content outside the drawer</Drawer>`

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+export interface DrawerProps extends React.HTMLProps<HTMLDivElement> {
+  /** Additional classes added to the Drawer. */
+  className?: string;
+  /** Content rendered inside the button */
+  children?: React.ReactNode;
+  /** Indicate if the drawer is expanded */
+  isExpanded: boolean;
+  /** Indicates if the content element and panel element are displayed side by side. */
+  isInline: boolean;
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.tsx]: should be optional?

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+import styles from '@patternfly/react-styles/css/components/Drawer/drawer';
+import { css } from '@patternfly/react-styles';
+
+export interface DrawerProps extends React.HTMLProps<HTMLDivElement> {
+  /** Additional classes added to the Drawer. */
+  className?: string;
+  /** Content rendered inside the button */
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.tsx]:
```suggestion
  /** Content rendered in the main section of the Drawer */
```

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  children?: React.ReactNode;
+  /** Indicate if the drawer is expanded */
+  isExpanded: boolean;
+  /** Indicates if the content element and panel element are displayed side by side. */
+  isInline: boolean;
+  /** Content to rendered in the left hand panel */
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.tsx]: I think the comments are opposite between this and `children`
```suggestion
  /** Content rendered in the drawer */
```

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+import styles from '@patternfly/react-styles/css/components/Drawer/drawer';
+import { css } from '@patternfly/react-styles';
+
+export interface DrawerPanelContentProps extends React.HTMLProps<HTMLDivElement> {
+  /** Additional classes added to the Drawer. */
+  className?: string;
+  /** Content to rendered in the left hand panel */
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx]:
```suggestion
  /** Content to rendered in the drawer */
```

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx (hunk 5 above)
Comment (dlabaj) [reply packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx]: okay will make this update

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md (hunk 6 above)
Comment (dlabaj) [reply packages/patternfly-4/react-core/src/experimental/components/Drawer/Drawer.md]: Done.

File: packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export const DrawerPanelContent: React.SFC<DrawerPanelContentProps> = ({
+  className = '',
+  children,
+  hasPadding = true,
+  ...props
+}: DrawerPanelContentProps) => (
+      <aside className={css(styles.drawerPanel, className)} {...props}>
+        <div className={css(styles.drawerPanelBody, hasPadding === false && styles.modifiers.noPadding)}>
```
Comment (mcoker) [inline packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx]: Should we use `noPadding`? We have this same feature in the wizard, and the prop is called `hasBodyPadding`. We also have it in the page component and data list and it's called `noPadding`. We support this feature in core in the table, too, but I looked in the react docs for the table component and couldn't find a reference to it, so not sure if we don't support it, or if we do and it's not documented.

And whichever direction we go, can you add it to the documentation?

Comment (dlabaj) [reply packages/patternfly-4/react-core/src/experimental/components/Drawer/DrawerPanelContent.tsx]: i'll change it to noPadding.

**Pull request merged**